### PR TITLE
submariner-diagnose: ignore diagnose pod on deployment test

### LINKED ISSUE #130: Error in running subctl diagnose from a pod - self pod appears to be 'pending'
**What happened**:
During subctl diagnose from inside a pod,
the test: checking sumariner pods test failed
✗ Pod ["submariner-diagnose-f4lcq" "Pending"] is not running. (current state is %!v(MISSING))

**What you expected to happen**:
The test was expected to succeed

**How to reproduce it (as minimally and precisely as possible)**:
Run subctl diagnose from a pod according to this manual:
https://submariner.io/operations/deployment/subctl/#running-subctl-diagnose-from-a-pod-in-cluster

**Anything else we need to know?**:

**Environment**:
- Diagnose information (use `subctl diagnose all`):
- Gather information (use `subctl gather`)
- Cloud provider or hardware configuration:
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by tomer-shefler**
Signed-off-by: Tomer Shefler <sheflertomer@gmail.com>

Fixes https://github.com/submariner-io/subctl/issues/130

### CODE REVIEW

**Comment by dfarrell07** [comment]
CC @mkolesnik as you were a part of the discussion that lead to this PR.

File: pkg/diagnose/deployments.go [hunk 1]
```diff
@@ -177,6 +176,9 @@ func checkPodsStatus(k8sClient kubernetes.Interface, namespace string, status re
 	for i := range pods.Items {
 		pod := &pods.Items[i]
 		if pod.Status.Phase != v1.PodRunning {
+			if pod.Labels["type"] == "submariner-diagnose" {
```
Comment (mkolesnik) [inline pkg/diagnose/deployments.go]: You should utilize label selectors in `ListOptions` instead (line 170).

You can find examples how to do it in the code.

Also I don't know if type is appropriate, maybe something like `transient: true` to indicate pods that we want to ignore (might be not just the diagnose pods in the future)

File: pkg/diagnose/deployments.go [hunk 2]
```diff
@@ -20,7 +20,6 @@ package diagnose
 
 import (
 	"context"
-
```
Comment (mkolesnik) [inline pkg/diagnose/deployments.go:23]: This change is unnecessary and also trips the go linter

**Review by mkolesnik — CHANGES_REQUESTED** [review]
Please update the commit message to explain **why** this change is necessary

**Comment by tomer-shefler** [comment]
Is dependent on https://github.com/submariner-io/subctl/issues/48

**Comment by mkolesnik** [comment]
Thanks, please follow guidelines for commit message: https://submariner.io/development/code-review/#commit-message-formatting

Also, as part of this fix you should add the label itself as part of diagnose pod deployment.

**Comment by tomer-shefler** [comment]
> Also, as part of this fix you should add the label itself as part of diagnose pod deployment.

This, was addded in here: https://github.com/submariner-io/submariner-website/pull/814
Is there anywhere else it sholud be added?

**Comment by mkolesnik** [comment]
> This, was addded in here: [submariner-io/submariner-website#814](https://github.com/submariner-io/submariner-website/pull/814) Is there anywhere else it sholud be added?

The diagnose FW command created the some pods, right?
So its a good idea to label them as such as well (to avoid causing any potential future problems).

File: pkg/diagnose/deployments.go [hunk 3]
```diff
@@ -172,7 +172,9 @@ func checkDaemonset(k8sClient kubernetes.Interface, namespace, daemonSetName str
 }
 
 func checkPodsStatus(k8sClient kubernetes.Interface, namespace string, status reporter.Interface) {
-	pods, err := k8sClient.CoreV1().Pods(namespace).List(context.TODO(), metav1.ListOptions{})
+	pods, err := k8sClient.CoreV1().Pods(namespace).List(context.TODO(), metav1.ListOptions{
+		LabelSelector: "transient!=true",
```
Comment (tpantelis) [inline pkg/diagnose/deployments.go]: We should qualify the label name to be safe.

```suggestion
		LabelSelector: "submariner.io/transient!=true",
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Comment by tomer-shefler** [comment]
> > This, was addded in here: [submariner-io/submariner-website#814](https://github.com/submariner-io/submariner-website/pull/814) Is there anywhere else it sholud be added?
> 
> The diagnose FW command created the some pods, right? So its a good idea to label them as such as well (to avoid causing any potential future problems).
They should be delete at the end of their test (when https://github.com/submariner-io/submariner-operator/pull/2168 is merged), so they won't bother this one. But I could label them either way.

File: pkg/diagnose/deployments.go [hunk 4]
```diff
@@ -172,7 +173,9 @@ func checkDaemonset(k8sClient kubernetes.Interface, namespace, daemonSetName str
 }
 
 func checkPodsStatus(k8sClient kubernetes.Interface, namespace string, status reporter.Interface) {
-	pods, err := k8sClient.CoreV1().Pods(namespace).List(context.TODO(), metav1.ListOptions{})
+	pods, err := k8sClient.CoreV1().Pods(namespace).List(context.TODO(), metav1.ListOptions{
+		LabelSelector: fmt.Sprint("%s!=%s", constants.TransientLabel, constants.TrueLabel),
```
Comment (mkolesnik) [inline pkg/diagnose/deployments.go]:
```suggestion
		LabelSelector: fmt.Sprintf("%s!=%s", constants.TransientLabel, constants.TrueLabel),
```

**Review by mkolesnik — CHANGES_REQUESTED** [review]

**Review by tpantelis — APPROVED** [review]

**Review by mkolesnik — APPROVED** [review]
Thanks!

**Comment by tpantelis** [comment]
I removed the dependency on https://github.com/submariner-io/submariner-operator/pull/2168 as it seems unrelated.

**Pull request merged**